docs/okeps/okep-5259-no-overlay.md

```diff
@@ -0,0 +1,1491 @@
+# OKEP-5259: No-overlay Mode For Layer-3 Networks using BGP
+
+* Issue: [\#5259](https://github.com/ovn-kubernetes/ovn-kubernetes/issues/5259)
+
+## Problem Statement
+
+Currently, OVN-Kubernetes uses Geneve as its encapsulation method on the overlay
+network for east-west traffic; this adds overhead and reduces throughput. By
+leveraging OVN-Kubernetes support for BGP, we want to provide a way for users to
+enable a no-overlay mode, which would disable Geneve encapsulation and use
+direct routing between nodes for east-west traffic on selected networks.
+
+Many environments, particularly on-premise deployments or those with dedicated
+networking infrastructure, prefer to utilize the underlying physical network's
+routing capabilities directly. This "no-overlay" approach can offer several
+benefits:
+
+* Improved Performance: Eliminates overlay protocol overhead, potentially
+  leading to lower latency and higher throughput for inter-pod communication.
+* Leverage Existing Network Infrastructure: Integrates more seamlessly with
+  existing BGP-capable network devices, allowing direct routing to pod IPs.
+* Reduced Resource Consumption: Fewer CPU cycles spent on
+  encapsulation/decapsulation.
+
+## Terminology
+
+See [terminology](https://ovn-kubernetes.io/okeps/okep-5193-user-defined-networks/#terminology) details.
+
+## Goals
+
+* Support no-overlay mode for the cluster default network (CDN).
+* Support no-overlay mode for Primary layer-3 ClusterUserDefinedNetworks
+  (CUDNs).
+* A cluster can have networks operating in overlay and no-overlay modes
+  simultaneously.
+* Use the BGP feature to exchange routes to pod subnets across the cluster.
+* Allow direct communication without any overlay encapsulation for east-west
+  traffic.
+* Maintain compatibility with existing OVN-Kubernetes features where applicable.
+* Compatible with both local gateway and shared gateway modes.
+* Do not compromise the existing advertised UDN isolation strict mode.
+
+## Future Goals
+
+* When OVN-Kubernetes supports BGP for UserDefinedNetwork CRs, extend no-overlay
+  mode support for Primary layer-3 UDNs.
+* Support toggling no-overlay mode on/off for an existing network. To support
+  this functionality, there are two things we need to consider.
+  First, in no-overlay mode, as there is no Geneve overhead, a network can have
+  the same MTU size as the provider network. However, overlay networks require a
+  smaller MTU (default 1400). We cannot modify a pod interface MTU without
+  restarting the pod, so OVN-Kubernetes must work with other external
+  controllers to accommodate such a migration.
+  Secondly, to facilitate a smooth, node-by-node migration, it's crucial to
+  allow networks to operate in a hybrid no-overlay and overlay mode.
+* Address the advertised UDN isolation loose mode issue when no-overlay mode is
+  enabled.
+
+## Non-Goals
```

Comment:
Not sure why we removed the non-goal of (not) supporting (tenant) UDNs

Reply:
It contradicts with future goal `When OVN-Kubernetes supports BGP for UserDefinedNetwork CRs, extend no-overlay mode support for Primary layer-3 UDNs.`